src/colorizer/TimeControls.ts

```diff
@@ -0,0 +1,132 @@
+import ColorizeCanvas from "./ColorizeCanvas";
+
+// time / playback controls
+export default class TimeControls {
+  private playBtn: HTMLButtonElement;
+  private pauseBtn: HTMLButtonElement;
+  private forwardBtn: HTMLButtonElement;
+  private backBtn: HTMLButtonElement;
+  private timeSlider: HTMLInputElement;
+  private timeInput: HTMLInputElement;
+
+  private timerId: number;
+  private redrawfn: () => void;
+
+  private canvas: ColorizeCanvas;
+  private isDisabled: boolean;
+
+  constructor(canvas: ColorizeCanvas, redrawfn: () => void) {
+    this.redrawfn = redrawfn;
+    this.canvas = canvas;
+    this.timerId = 0;
+    this.isDisabled = false;
+
+    this.playBtn = document.querySelector("#playBtn")!;
+    this.pauseBtn = document.querySelector("#pauseBtn")!;
+    this.forwardBtn = document.querySelector("#forwardBtn")!;
+    this.backBtn = document.querySelector("#backBtn")!;
+    this.timeSlider = document.querySelector("#timeSlider")!;
+    this.timeInput = document.querySelector("#timeValue")!;
+    this.playBtn.addEventListener("click", () => this.handlePlayButtonClick());
+    this.pauseBtn.addEventListener("click", () => this.handlePauseButtonClick());
+    this.forwardBtn.addEventListener("click", () => this.handleFrameAdvance(1));
+    this.backBtn.addEventListener("click", () => this.handleFrameAdvance(-1));
+    // only update when DONE sliding: change event
+    this.timeSlider.addEventListener("change", () => this.handleTimeSliderChange());
+    this.timeInput.addEventListener("change", () => this.handleTimeInputChange());
+  }
+
+  private wrapFrame(index: number): number {
+    const totalFrames = this.canvas.getTotalFrames();
+    return (index + totalFrames) % totalFrames;
+  }
+  
+  private clampFrame(index: number): number {
+    return Math.min(Math.max(index, 0), this.canvas.getTotalFrames() - 1);
+  }
+
+  private playTimeSeries(onNewFrameCallback: () => void): void {
+    clearInterval(this.timerId);
+
+    // TODO: Fix this function so that it doesn't stop the
+    // slider from also operating
+    const loadNextFrame = async (): Promise<void> => {
+      if (this.isDisabled) {
+        // stop render loop if time controls have been disabled.
+        return;
+      }
+      let nextFrame = this.canvas.getCurrentFrame() + 1;
+      if (nextFrame >= this.canvas.getTotalFrames()) {
+        nextFrame = 0;
+      }
```

Comment:
could now use
`let nextFrame = this.wrapFrame(this.canvas.getCurrentFrame() + 1);`